plugins/Synchro/SynchroSynth.cpp

```diff
@@ -0,0 +1,389 @@
+/*
+ * SynchroSynth.cpp - 2-oscillator PM synth
+ *
+ * Copyright (c) 2019 Ian Sannar <ian/dot/sannar/at/gmail/dot/com>
+ * Credit to @DouglasDGI "Lost Robot" for performance optimizations
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "SynchroSynth.h"
+//Standard headers
+#include <math.h>
+//QT headers
+#include <QDomDocument>
+//LMMS headers
+#include "embed.h"
+#include "Engine.h"
+#include "InstrumentTrack.h"
+#include "lmms_constants.h"
+#include "Mixer.h"
+#include "NotePlayHandle.h"
+#include "plugin_export.h"
+
+#define SYNCRHO_VERSION "0.4"
+#define SYNCHRO_VOLUME_CONST 0.15f
+
+static const int SYNCHRO_OVERSAMPLE = 4; //Anti-aliasing samples
+static const int SYNCHRO_PM_CONST = 20; //Strength of the phase modulation
+static const int SYNCHRO_GRAPH_SAMPLES = 48; //Resolution of the waveform view GUI elements
+
+//Plugin metadata
+extern "C"
+{
+	Plugin::Descriptor PLUGIN_EXPORT synchro_plugin_descriptor =
+	{
+		STRINGIFY(PLUGIN_NAME),
+		"Synchro",
+		QT_TRANSLATE_NOOP("pluginBrowser", "2-oscillator PM synth"),
+		"Ian Sannar <ian/dot/sannar/at/gmail/dot/com>",
+		0x0100, //What does this do?
+		Plugin::Instrument,
+		new PluginPixmapLoader("logo"),
+		NULL,
+		NULL
+	};
+	// neccessary for getting instance out of shared lib
+	PLUGIN_EXPORT Plugin * lmms_plugin_main(Model * m, void * data)
+	{
+		return new SynchroSynth(static_cast<InstrumentTrack *>(m));
+	}
+}
+
+SynchroNote::SynchroNote(NotePlayHandle * nph) :
+	nph(nph)
+{
+	//Empty constructor
+}
+
+SynchroNote::~SynchroNote()
+{
+	//Empty destructor
+}
+
+void SynchroNote::nextStringSample(sampleFrame &outputSample,
+	sample_rate_t sample_rate, float modulationStrength,
+	float modulationAmount, float harmonics, SynchroOscillatorSettings carrier,
+	SynchroOscillatorSettings modulator)
+{
+	float sampleRatePi = F_2PI / sample_rate; //For oversampling
+	//Find position in modulator waveform
+	sample_index[1] += sampleRatePi * DetuneOctaves(nph->frequency(),
+		modulator.Detune);
+	while (sample_index[1] >= F_2PI) { sample_index[1] -= F_2PI; } //Make sure phase is always between 0 and 2PI
```

Comment:
Is `fmod` faster? I'm open to using that instead. I think that would maintain readability while making things more concise.

Reply:
It would be difficult to say without profiling it, but I don't think it will have a significant impact. `fmod` is inherently slower than subtraction, but the while loop may need to do multiple comparisons and subtractions, and is also subject to branch prediction. `fmod` will at least be more consistent in speed. I would be surprised if this were a bottleneck though.

One nice advantage to `fmod` is that if some disaster occurs and `sample_index` becomes infinite (or large enough that subtracting `F_2PI` makes no difference), then `fmod` will return `NaN` (or some value at least), whereas the while loop will hang. It might sound unlikely, but I've seen it happen a few times in otherwise correct code.

My main motivation for this suggestion though is that I think it's easier to read, and clearer that it's not going to misbehave. That might just be personal preference though, so I'll leave it up to you.

(One thing to remember is that `fmod` returns a negative value for negative inputs, so you'll need to write `if (sample_index[0] < 0) { sample_index[0] += F_2PI; }` if you choose to use it in the second instance lower down. This can be an if rather than a while though.)